### ISSUE TITLE
Can we decrease gap communication times?

### ISSUE THREAD

**Issue by lindsayad** [idaholab/moose#22994]
## Reason
Here is a difference in gap vs. velocity communication time:
```
(pprof) list incorrectEdgeDroppingPost
Total: 197.98s
ROUTINE ======================== ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost in /home/lindad/projects/moose/modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C
         0    100.30s (flat, cum) 50.66% of Total
         .          .    257:void
         .          .    258:ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost(
         .          .    259:    const std::unordered_set<const Node *> & inactive_lm_nodes)
         .          .    260:{
         .          .    261:#ifdef MOOSE_SPARSE_AD
         .    100.29s    262:  Moose::Mortar::Contact::communicateGaps(
         .          .    263:      _dof_to_weighted_gap, this->processor_id(), _mesh, _nodal, _normalize_c, _communicator);
         .       10ms    264:  Moose::Mortar::Contact::communicateVelocities(
         .          .    265:      _dof_to_weighted_tangential_velocity, this->processor_id(), _mesh, _nodal, _communicator);
         .          .    266:  Moose::Mortar::Contact::communicateVelocities(
         .          .    267:      _dof_to_real_tangential_velocity, this->processor_id(), _mesh, _nodal, _communicator);
         .          .    268:#endif
         .          .    269:
```
Per the graph below this is 50% of the simulation time
![threed-mortar-strumpack-again0](https://user-images.githubusercontent.com/10248304/208546834-1bfe3b5b-fc5f-4d11-9f06-1d32c2db1ac2.svg)

## Design
Not sure yet

## Impact
Speed up mortar contact

**Comment by lindsayad** [comment]
@roystgnr the above data is for a simulation with 12 processes. If I print the communication pattern like this (and similarly for velocities):
```c++
  for (const auto & [pid, vec] : push_data)
    std::cout << "From " << communicator.rank() << ", communcating gaps. Pushing to process " << pid
              << " a vector of length " << vec.size() << std::endl;
  TIMPI::push_parallel_vector_data(communicator, push_data, action_functor);
```
then I get a printout like this:
```
From 10, communcating gaps. Pushing to process 9 a vector of length 9
From 10, communcating velocities. Pushing to process 9 a vector of length 9
From 10, communcating velocities. Pushing to process 9 a vector of length 9
From 10, communcating gaps. Pushing to process 9 a vector of length 9
From 10, communcating velocities. Pushing to process 9 a vector of length 9
From 10, communcating velocities. Pushing to process 9 a vector of length 9
```
I am actually calling push twice as much for the velocities, but communicating gaps takes 100 seconds while communicating velocities takes 10ms. Any ideas?

**Comment by lindsayad** [comment]
I'm a dummy. If I swap calls to `communicateGaps` and `communicateVelocities`, then all the time is spent in `communicateVelocities` ... and this is because of bad load balancing! On process 9, things look perfectly normal because that process has a bunch of mortar mesh computations. The problem is that only process 9 and 10 have mortar segments, so 10/12 processes do nothing for 50% of the simulation 😆 

![threed-mortar-strumpack-again9](https://user-images.githubusercontent.com/10248304/208569650-34691c71-a177-490a-bdbc-9b357c8d8bcc.svg)

**Comment by recuero** [comment]
> This disparity is https://github.com/idaholab/moose/issues/22994#issuecomment-1358765027.

Can we weigh the domain decomposition by the sidesets that generate the mortar interface?

**Comment by lindsayad** [comment]
Yes I think we could definitely do something like that. I think I'd have to talk with @roystgnr about how to do that (or spend some time looking at the `Partitioner` code)

**Comment by roystgnr** [comment]
At the moment we have a base class API for weighting the graph to be partitioned, but it's only implemented in MetisPartitioner, and it only lets us weight graph vertices (corresponding to Elems), not graph edges.  So we can tell MetisPartitioner "this element makes a lot of extra work for us, don't put as many other elements on the rank which ends up owning it", but we can't tell it "these two elements really need a ton of extra communication with each other, try extra hard to keep them owned by the same rank".

Adding ParmetisPartitioner support for Elem weights would be pretty trivial.  Adding support for edge weights ... looks like that corresponds to the "adjacency weights" option in Metis and Parmetis?  So not trivial, but the hard work seems to have already been done.

**Comment by lindsayad** [comment]
The beginning of this conversation is probably misleading. We do want to weight graph vertices. The current issue (the original proposed issue of communication was a red herring) is that two processes do all the mortar work and ten processes just sit and wait. So we’d want to use the already existing APIs for weighting vertices and distributing those mortar elements around
